Scijava/scijava ops engine/skip map history

hinerm opened the pull request:
Closes https://github.com/scijava/scijava/issues/99

See also [#100](https://github.com/scijava/scijava/issues/100) and [#101](https://github.com/scijava/scijava/issues/101)

hinerm committed:
RichOp: skip output logging if hint is present

History.SKIP_RECORDING hint is used to skip history recording in the
AbstractRichOp constructor, so this hint should also be respected in the
postprocess step of recording output.

hinerm committed:
MapNeighborhood: fix indent

hinerm committed:
DefaultMapNeighborhood: rename variable

This isn't the center-aware impl

hinerm committed:
MapNeighborhood: disable per-pixel history

Per-pixel operations with imglib2 we can reasonably assume to be
operating on Type wrappers that have some level of reuse. These are not
thread-safe to use as output keys in our history, thus we cannot be
recording per-pixel history.

hinerm committed:
Do not force History.SKIP_RECORDING

The ProvenanceTest relies on history being recorded and this always
skipping seems aggressive.

ctrueden commented on imagej/imagej-ops2/src/main/java/net/imagej/ops2/map/neighborhood/MapNeighborhoodWithCenter.java:
```diff
@@ -118,10 +120,20 @@ public void compute(final RandomAccessibleInterval<I> in1, final Shape in2,
+
+		if (centerAwareOp instanceof RichOp) {
+			((RichOp)centerAwareOp).metadata().hints().minus(
+					BaseOpHints.History.SKIP_RECORDING);
```
If `SKIP_RECORDING` hint was already set before getting plussed in above... then should we still be minusing it out again here? Don't we just want to restore the hints back to their original state?

ctrueden reviewed (approved):
Looks awesome. I commented on one small thing, but I think we should merge this as-is as soon as possible, to get the nondeterministic failures under control.

ctrueden commented:
On this branch, I do get one failure in imagej-ops2, on both the CLI and in IntelliJ:
```
[ERROR] Tests run: 3, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 0.755 s <<< FAILURE! - in net.imagej.ops2.features.tamura2d.Tamura2dFeatureTest
[ERROR] testCoarsenessFeature  Time elapsed: 0.639 s  <<< ERROR!
java.lang.ArrayIndexOutOfBoundsException
	at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.features.tamura2d.Tamura2dFeatureTest.testCoarsenessFeature(Tamura2dFeatureTest.java:65)
Caused by: java.lang.ArrayIndexOutOfBoundsException: Index 10000 out of bounds for length 10000
```

According to the stack trace, it's still trying to log the output of individual pixels. It's like the hint isn't being respected.

ctrueden reviewed (changes requested):
On second thought, I think we better get to the bottom of the hint-propagation failure before merging this. I strongly considered just commenting out the offending test, but it's going to bite us sooner rather than later if we do that.

hinerm commented:
> On second thought, I think we better get to the bottom of the hint-propagation failure before merging this. I strongly considered just commenting out the offending test, but it's going to bite us sooner rather than later if we do that.

@ctrueden see https://github.com/scijava/scijava/issues/101#issuecomment-1289343895

hinerm committed:
OpMetadata: allow updating of Hints

There are cases we want to control what the Hints are during some stages
of matching.

hinerm commented:
@ctrueden @gselzer I just added a commit allowing the updating of `Hints` in an `OpMetadata`.. these tests are passing consistently for me now. Can you test a few times on your end and see if you get any failures?

ctrueden reviewed (approved):
<details><summary>With OpenJDK 17.0.3, there is an unrelated JPMS issue with scijava-persist</summary>

```
[INFO] Running org.scijava.persist.SerializationTests
IClassAdapters : 
IRunTimeClassAdapters : 
	 interface org.scijava.persist.Shape
	 	 class org.scijava.persist.Circle
	 	 class org.scijava.persist.Shapes$Diamond
	 	 class org.scijava.persist.Shapes$Rectangle
IClassAdapters : 
IRunTimeClassAdapters : 
	 interface org.scijava.persist.Shape
	 	 class org.scijava.persist.Circle
	 	 class org.scijava.persist.Shapes$Diamond
	 	 class org.scijava.persist.Shapes$Rectangle
[ERROR] Tests run: 2, Failures: 0, Errors: 2, Skipped: 0, Time elapsed: 0.213 s <<< FAILURE! - in org.scijava.persist.SerializationTests
[ERROR] testCircleObject  Time elapsed: 0.192 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field int org.scijava.persist.Circle.radius accessible: module org.scijava.persist does not "opens org.scijava.persist" to module com.google.gson
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testSerializationDeserialization(SerializationTests.java:76)
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testCircleObject(SerializationTests.java:50)

[ERROR] testDrawingObject  Time elapsed: 0.012 s  <<< ERROR!
java.lang.reflect.InaccessibleObjectException: Unable to make field int org.scijava.persist.Circle.radius accessible: module org.scijava.persist does not "opens org.scijava.persist" to module com.google.gson
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testSerializationDeserialization(SerializationTests.java:76)
	at org.scijava.persist@0-SNAPSHOT/org.scijava.persist.SerializationTests.testDrawingObject(SerializationTests.java:62)
```

</details>

With OpenJDK 11.0.16, I got the following results:

1. BUILD SUCCESS on Ubuntu
2. BUILD SUCCESS on macOS
3. BUILD FAILURE on Ubuntu: `org.scijava.ops.api.features.OpMatchingException: No MatchingRoutine was able to produce a match! at net.imagej.ops2@0-SNAPSHOT/net.imagej.ops2.identity.IdentityTest.testIdentity(IdentityTest.java:49)`
4. BUILD SUCCESS on macOS
5. BUILD SUCCESS on Ubuntu

This `No MatchingRoutine was able to produce a match!` error is not new; it was happening intermittently previously, and will need investigation, but I do not think it should hold up merging of this PR, which does seem to address the AIOOBEs and likely performance issues of map operations.
